airflow/dag_fetcher.py

```diff
@@ -0,0 +1,290 @@
+        """
+        if os.path.isfile(self.dags_uri):
+            self.process_file(self.dags_uri, only_if_updated=only_if_updated)
+        elif os.path.isdir(self.dags_uri):
+            patterns = []
+            for root, dirs, files in os.walk(self.dags_uri, followlinks=True):
+                ignore_file = [f for f in files if f == '.airflowignore']
+                if ignore_file:
+                    f = open(os.path.join(root, ignore_file[0]), 'r')
+                    patterns += [p for p in f.read().split('\n') if p]
+                    f.close()
+                for f in files:
+                    try:
+                        filepath = os.path.join(root, f)
+                        if not os.path.isfile(filepath):
+                            continue
+                        mod_name, file_ext = os.path.splitext(
+                            os.path.split(filepath)[-1])
+                        if file_ext != '.py' and not zipfile.is_zipfile(filepath):
+                            continue
+                        if not any(
+                                [re.findall(p, filepath) for p in patterns]):
+                            ts = timezone.utcnow()
+                            found_dags = self.process_file(
+                                filepath, only_if_updated=only_if_updated)
+
+                            td = timezone.utcnow() - ts
+                            td = td.total_seconds() + (
+                                float(td.microseconds) / 1000000)
+                            self.stats.append(self.FileLoadStat(
+                                filepath.replace(self.dags_uri, ''),
+                                td,
+                                len(found_dags),
+                                sum([len(dag.tasks) for dag in found_dags]),
+                                str([dag.dag_id for dag in found_dags]),
+                            ))
+                    except Exception as e:
+                        self.log.exception(e)
+
+        return self.stats
+
+
+class HDFSDagFetcher(DagFetcher):
+    """
+    HDFSDagFetecher - Not Implemented
+    """
+
+
+class S3DagFetcher(DagFetcher):
+    """
+    S3DagFetcher - Not Implemented
+    """
+
+
+class GCSDagFetcher(DagFetcher):
+    """
+    GCSDagFetcher - Not Implemented
+    """
+
+
+class GitRepoDagFetcher(DagFetcher):
+    """
+    GitRepoDagFetcher - Not Implemented
+    """
+
+
+def get_dag_fetcher(dagbag, dags_uri):
+    """
+    Factory method that returns an instance of the right
+    DagFetcher, based on the dags_uri prefix.
+
+    Any prefix that does not match keys in the dag_fetchers
+    dict (or no prefix at all) defaults to FileSystemDagFetcher.
+    """
+    log = LoggingMixin().log
+
+    dag_fetchers = dict(
+        hdfs=HDFSDagFetcher,
+        s3=S3DagFetcher,
+        gcs=GCSDagFetcher,
+        git=GitRepoDagFetcher)
+
+    uri_schema = dags_uri.split(':')[0]
```

Comment:
this is a bit rigid, it doesn't handle scenario where multiple dag fetchers are written for a specific service

Reply:
I think it maybe worth adding dag fetcher as a default config.